Refactor our "Send to Phone" feature using a serverless function

sraby opened the pull request:
This PR refactors our "Send to Phone" feature on the court results page to make use of our new Netlify serverless functions infrastructure. In doing so, this PR adds a few more env vars for configuration and also removes the need for an APIClient file, as our front-end now only communicates with our own defined functions.

toolness commented on src/components/SaveToPhone.js:
```diff
@@ -24,20 +30,16 @@ class SaveToPhone extends React.Component {
     // Needed for gatsby builds that don't have references to window
     if (typeof window !== "undefined") {
-      page = window.location.href;
+      path = window.location.pathname;
```
just so you know, I don't think this will include the `?` in a href and anything after it, so e.g. in `/en-US/questions?step=1` it won't include the `?step=1`.  So if you want the whole thing you'll likely want to do `path = window.location.pathname + window.location.search`.

sraby replied:
Ahhh good point. I don't think those results pages ever have params like that but we should make sure we can support if we ever do.

toolness commented on src/components/SaveToPhone.js:
```diff
@@ -24,20 +30,16 @@ class SaveToPhone extends React.Component {
-    APIClient.sendSMS(this.state.phone, message).then((res) => {
-      if (res.statusCode >= 400) {
+    sendSMS(this.state.phone, path).then((res) => {
+      if (res.status >= 400) {
```
Since this is the result of a `fetch` you might want to use the [`ok` property](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok) (i.e. `if (res.ok)`) here instead, it's a bit more readable.  It also will be `false` if the result was a redirect (HTTP status code in the 300-399 range), which it never should be.

sraby replied:
Noice

toolness commented on src/serverless-functions/send-to-phone.ts:
```diff
@@ -0,0 +1,43 @@
+  const client = new Twilio(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN);
+
+  await client.messages.create({
+    body: `Eviction Free NYC! Follow this link for assistance in your eviction case: https://${EFNYC_HOST}${
+      resultsPagePath || ""
+    }`,
```
since you're not doing any validation of `resultsPagePath`, this actually has a bit of a security hole in it: an attacker could set up a domain like `www.evicitonfreenyc.org.evil.com` and pass a `resultsPagePath` of `.evil.com/somepage` to send the user to a domain they control!  So you might want to do some basic validation of `resultsPagePath`, e.g. test that it starts with a `/` at least.

sraby replied:
Ahhhh wow didn't even think of that!

sraby replied:
Also this attacker must be a pretty big deal if they own the domain `evil.com`...

toolness commented on src/serverless-functions/send-to-phone.ts:
```diff
@@ -0,0 +1,43 @@
+  const resultsPagePath =
+    event.queryStringParameters && event.queryStringParameters.userPath;
+
+  if (!userPhone) {
+    return {
+      statusCode: 400,
+      body: "No receiving phone number given!",
+    };
+  }
```
It might be a good idea to validate the phone number somewhat here, just to be safe.  I don't think anything especially horrible will happen if we give twilio a totally crazy phone number here but it's good to follow the [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle).

sraby replied:
Yep, good point.

toolness commented on src/serverless-functions/send-to-phone.ts:
```diff
@@ -0,0 +1,43 @@
+    throw new Error("Twilio credentials are not in environment!");
+  }
+
+  const userPhone =
+    event.queryStringParameters && event.queryStringParameters.userPhone;
+
+  const resultsPagePath =
+    event.queryStringParameters && event.queryStringParameters.userPath;
```
It might be good to use `event.body` here instead of `event.queryStringParameters`, making this method something that's used with a POST request instead of a GET, since the function effectively has side effects (in that it sends an SMS message to someone).  You'd then want to change the `sendSMS` function to issue a POST request with the arguments in the body of the request too.

sraby replied:
Cool cool

toolness reviewed (commented):
Nooice this is looking good!  A few comments below.

sraby commented:
Alright @toolness, I addressed your review comments and also converted our sendSMS client function to a POST request. If you don't mind giving this a quick look over once again, I would appreciate it!

toolness commented on src/serverless-functions/send-to-phone.ts:
```diff
@@ -0,0 +1,67 @@
+    throw new Error("Twilio credentials are not in environment!");
+  }
+
+  const userPhone = event.body && JSON.parse(event.body).userPhone;
```
One thing to keep in mind is that the event body might not actually be JSON, so this might throw an exception. Ideally we should return a 400 (bad request) or something similar in such a case, rather than a 500 internal server error, but bleh whatever.

Also keep in mind that JSON can be any kind of object!  It could be just `null`, in which case the `JSON.parse()` will succeed but getting `userPhone` will throw an exception, or `userPhone` could be something other than a string, etc.  If this were an API that lots of different clients were using, we might want to use something like [JSON schema](https://json-schema.org/) to make sure the deserialized object had the right shape but here it's probably not a big deal.

(Actually this is sort of a decent reason for only accepting `application/x-www-form-urlencoded` payloads, since that only supports string keys and values, so it means there's way less validation to do, but ehh don't worry about it here.)

toolness commented on src/components/SaveToPhone.js:
```diff
@@ -8,6 +7,23 @@ import ReactPhoneInput from "react-phone-input-2";
+  };
+  return fetch(url, {
+    headers: {
+      Accept: "application/json",
+      "Content-Type": "application/json",
```
o nice thanks for adding these.

toolness commented on src/components/SaveToPhone.js:
```diff
@@ -8,6 +7,23 @@ import ReactPhoneInput from "react-phone-input-2";
+      "Content-Type": "application/json",
+    },
+    method: "POST",
+    mode: "cors",
```
Hmm I don't _think_ you need `mode: "cors"` here since it's a request to the same server?

sraby replied:
right right

toolness reviewed (approved):
Noooice this looks good!  The one last thing I might consider is writing a few unit tests, which could be useful for at least some of your validation functions.  You could also unit test the whole handler (see the [hello world test](https://github.com/JustFixNYC/eviction-free-nyc/blob/master/src/serverless-functions/hello-world.test.ts) for an example) but you'd likely have to mock out the twilio functionality which could be annoying.

sraby commented:
@toolness Just filed #128 to address your testing rec!
